### PR TITLE
Add the service to run jobs in the background thread pool instead of "execution phase started" callback.

### PR DESCRIPTION
**Pull request by mlopatkin**
This is an alternative approach to solve "background job adds unexpected configuration inputs" problem. Instead of having a gatekeeping callback, a service to run arbitrary Callables in background is provided. The code of the Callables will be exempt from the configuration input tracking (though not in this PR).

### CODE REVIEW

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (wolfs) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:34]: Why is this a `Callable` and not a `Runnable` if it can't return anything to the caller?

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (wolfs) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:39]: What is this used for?

**Review by wolfs — COMMENTED** [review]
I left two questions. I am also probably missing some context.

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (mlopatkin) [reply subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:39]: It is used by the scan plugin internally, to figure out if one background task is submitting another background task.

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy
Comment (adammurdoch) [inline subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy:66]: I wonder whether these should also fail the build. I guess the idea is that the scan plugin collects the failures and makes its own decision whether to fail or not? In which case, the scan plugin should maybe pass a `Callable` that does the collecting, and the contract of `execute(Callable)` would be to fail the build if the exception propagates out of the callable. We can also get rid of the callback in this case too.

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy
Comment (adammurdoch) [inline subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy:81]: I'd probably check somewhere that the jobs actually do run in the background and do not block the task from finishing, and also that multiple jobs can run in parallel.

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (adammurdoch) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:45]: This seems like something Gradle should take care of, rather than the scan plugin.

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (adammurdoch) [reply subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:39]: Why does the scan plugin care that a background task is submitting another task? Does it add some validation to prevent this? If so, it's probably something that Gradle should take care of (either by supporting it or preventing it from happening).

**Review by adammurdoch — COMMENTED** [review]
Looks good, I've added a few comments for things we might tweak.

**Comment by mlopatkin** [comment]
@bot-gradle test this

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (ldaley) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:33]: From my naive and context-lacking perspective, this construct looks overkill. 

Might it be better to have `GradleEnterprisePluginBackgroundJobExecutor` not is-a `Executor` but instead `has-a` two different `Executor`s that behave with the traits? That would cut down the API surface quite a bit.

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy
Comment (ldaley) [inline subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy:133]: I'd like to see more of what the user experience is here for this kind of failure.

**Review by ldaley — CHANGES_REQUESTED** [review]

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (ldaley) [reply subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:33]: I think it would be better to avoid the implementation details here and expose more “logical” executors with the right behaviour. 

Along those lines, could that be handled by the Gradle runtime? Gradle knows what the thread is doing when background work is submitted. It could use that context to do the appropriate decoration.

Comment (mlopatkin) [reply subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:33]: > I think it would be better to avoid the implementation details here and expose more “logical” executors with the right behaviour.

Thanks, I see your point. This also opens some leeway in changing the properties of the underlying executor. I guess I shouldn't attempt to overgeneralize this API, especially considering it is an interim solution.

> Along those lines, could that be handled by the Gradle runtime? Gradle knows what the thread is doing when background work is submitted. It could use that context to do the appropriate decoration.

I'm not sure if I fully understand your idea here. Is it about deducing the traits automatically from the context? I think this couples the plugin and the core a bit too much.

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy
Comment (adammurdoch) [inline subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy:153]: We'd usually use `BlockingHttpServer` instead of latches and timeouts to test parallel execution from an int test.

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy
Comment (adammurdoch) [inline subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy:202]: Same here, we'd usually use `BlockingHttpServer` for this kind of test.

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy
Comment (adammurdoch) [inline subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy:321]: It would be better to use the `ConfigurationCacheFixture` rather than hard-coded console output to check for loads and stores.

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java
Comment (adammurdoch) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutor.java:29]: It might be better to has-a Executor than is-a Executor, for example so that we can later offer different executors with different semantics, or even not use an Executor at all.

File: subprojects/base-services/src/main/java/org/gradle/internal/concurrent/ManagedExecutorImpl.java
Comment (bamboo) [inline subprojects/base-services/src/main/java/org/gradle/internal/concurrent/ManagedExecutorImpl.java:39]: 💄 Consider annotating the package with `@NonNullApi` instead.

File: subprojects/composite-builds/src/test/groovy/org/gradle/composite/internal/DefaultIncludedBuildRegistryTest.groovy
Comment (bamboo) [inline subprojects/composite-builds/src/test/groovy/org/gradle/composite/internal/DefaultIncludedBuildRegistryTest.groovy:67]: 💄 Please keep the multiline format for very long lines like this one as it makes it easier to review the code.

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginBackgroundJobExecutor.java
Comment (bamboo) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginBackgroundJobExecutor.java:55]: Not a strong feeling but I'd move this function to `Instrumented` and hide the individual `disable` / `restore` steps completely.

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginBackgroundJobExecutor.java
Comment (bamboo) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginBackgroundJobExecutor.java:66]: Cool trick 👍

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy
Comment (mlopatkin) [reply subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutorIntegrationTest.groovy]: Done, thanks for pointing to this class

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutors.java
Comment (bamboo) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/GradleEnterprisePluginBackgroundJobExecutors.java:43]: 👍

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginBackgroundJobExecutors.java
Comment (bamboo) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginBackgroundJobExecutors.java:53]: 👍

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginRequiredServices.java
Comment (bamboo) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginRequiredServices.java:30]: 💄 Rename field to `backgroundJobExecutors` to match the type name.

**Review by bamboo — APPROVED** [review]

**Comment by mlopatkin** [comment]
After a discussion within the CC team, we think that this API looks shippable. We don't have an explicit approval from @ldaley yet, but don't want to block 7.6 RC1 release too. I'm going to merge this as is to unblock 7.6. If any tweaks are required, we can do them in separate PRs targeting 7.6 RC2/7.6.1/8.0, then the scan plugin can start using the new API in the version that includes all necessary tweaks.

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Pull request merged**
